### PR TITLE
Delay background jobs execution

### LINKED ISSUE #5892: Delay Background Jobs Execution
This task is a part of building a warmup infrastructure #5040

## Motivation

The engine runs a queue of background jobs. During the application startup, these jobs can be delayed until the first program execution is complete. This way they are not interfering with the execution job and are not competing for available CPU resources.

## Design

- Delay the engine background jobs until the first execution is complete
- Send a notification when the background jobs are started

### PR DESCRIPTION
**Pull request by 4e6**
### Pull Request Description

close #5892

Changelog:
add: feature to delay background jobs execution
add: start background jobs when program finishes
add: start background jobs on `search/completion` request

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] ~If GUI codebase was changed, the GUI was tested when built using `./run ide build`.~

### CODE REVIEW

File: engine/polyglot-api/src/main/scala/org/enso/polyglot/runtime/Runtime.scala [hunk 1]
```diff
@@ -1665,6 +1673,12 @@ object Runtime {
     final case class DeserializeLibrarySuggestions(libraryName: LibraryName)
         extends ApiRequest
 
+    /** A request to start the background jobs processing. */
```
Comment (JaroslavTulach) [inline engine/polyglot-api/src/main/scala/org/enso/polyglot/runtime/Runtime.scala:1676]: The comment could mention who emits the message and when.

File: engine/language-server/src/main/scala/org/enso/languageserver/search/SuggestionsHandler.scala [hunk 2]
```diff
@@ -385,6 +385,17 @@ final class SuggestionsHandler(
           )
         }
         .pipeTo(sender())
+      if (state.shouldStartBackgroundProcessing) {
+        runtimeConnector ! Api.Request(Api.StartBackgroundProcessing())
```
Comment (JaroslavTulach) [inline engine/language-server/src/main/scala/org/enso/languageserver/search/SuggestionsHandler.scala:389]: Does this mean that `SuggestionHandler` emits the `Api.StartBackgroundProcessing` message? If so, it is strange. I'd expect some more central authority to emit such message.

File: engine/runtime/src/main/scala/org/enso/interpreter/instrument/command/DeserializeLibrarySuggestionsCmd.scala [hunk 3]
```diff
@@ -19,8 +19,10 @@ class DeserializeLibrarySuggestionsCmd(
   override def execute(implicit
     ctx: RuntimeContext,
     ec: ExecutionContext
-  ): Future[Unit] =
+  ): Future[Unit] = {
     ctx.jobProcessor.runBackground(
       new DeserializeLibrarySuggestionsJob(request.libraryName)
     )
+    Future.successful(())
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/scala/org/enso/interpreter/instrument/command/DeserializeLibrarySuggestionsCmd.scala:26]: Is the operation `successful` when the job is just submitted? Shouldn't `runBackground` return a `Future` to track how "successful" the operation is?

File: engine/runtime/src/main/scala/org/enso/interpreter/instrument/command/StartBackgroundProcessingCmd.scala [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  maybeRequestId: Option[Api.RequestId]
+) extends Command(maybeRequestId) {
+
+  /** @inheritdoc */
+  override def execute(implicit
+    ctx: RuntimeContext,
+    ec: ExecutionContext
+  ): Future[Unit] = {
+    StartBackgroundProcessingJob.startBackgroundJobs()
+    Future.successful(())
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/scala/org/enso/interpreter/instrument/command/StartBackgroundProcessingCmd.scala:23]: Returning `successful` here is probably OK - `StartBackgroundProcessingJob` is probably successful when the background processing got started.

File: engine/polyglot-api/src/main/scala/org/enso/polyglot/runtime/Runtime.scala [hunk 5]
```diff
@@ -1665,6 +1673,12 @@ object Runtime {
...
+    final case class StartBackgroundProcessing() extends ApiRequest
+
+    /** A notification about started background jobs. */
+    final case class BackgroundJobsStartedNotification() extends ApiNotification
```
Comment (JaroslavTulach) [inline engine/polyglot-api/src/main/scala/org/enso/polyglot/runtime/Runtime.scala:1680]: What is this `BackgroundJobsStartedNotification` good for?

**Review by JaroslavTulach — APPROVED** [review]
Who emits the `StartBackgroundProcessing` is my main concern. Otherwise seems OK.

File: engine/language-server/src/main/scala/org/enso/languageserver/search/SuggestionsHandler.scala (hunk 2 above)
Comment (4e6) [reply engine/language-server/src/main/scala/org/enso/languageserver/search/SuggestionsHandler.scala:389]: `StartBackgroundProcessing` is just a command that triggers the warmup logic. The command handler is responsible for emitting the `BackgrouncProcessingStarted` notification.

File: engine/runtime/src/main/scala/org/enso/interpreter/instrument/command/DeserializeLibrarySuggestionsCmd.scala (hunk 3 above)
Comment (4e6) [reply engine/runtime/src/main/scala/org/enso/interpreter/instrument/command/DeserializeLibrarySuggestionsCmd.scala:26]: `runBackground` just schedules the job execution and does not imply that the job will be started (or even submitted) for execution right away.

File: engine/polyglot-api/src/main/scala/org/enso/polyglot/runtime/Runtime.scala (hunk 5 above)
Comment (4e6) [reply engine/polyglot-api/src/main/scala/org/enso/polyglot/runtime/Runtime.scala:1680]: It is a notification sent when the engine starts the background jobs processing. It is different from the `StartBackgoundProcessing` which is a command to start the background jobs processing

Comment (4e6) [reply engine/polyglot-api/src/main/scala/org/enso/polyglot/runtime/Runtime.scala:1680]: It is a notification sent when the engine starts the background jobs processing. It is different from the `StartBackgoundProcessing` which is a command to start the background jobs processing

**Pull request merged**